### PR TITLE
Controller#request2lang is coupled with the i18n API that relies on runtime DI

### PR DESCRIPTION
**Pull request by julienrf**
See [here](https://github.com/playframework/playframework/blob/master/framework/src/play/src/main/scala/play/api/mvc/Controller.scala#L65). It means that you get a runtime exception if you try to get an implicit `Lang` from within an action and if your code does not use a runtime DI mechanism.

### CODE REVIEW

**Comment by julienrf** [comment]
I attached a PR to the issue. The PR also fixes #3781.

Please give me some feedback on the PR. I didn’t updated the documentation yet. My ideas are the following:
- define a healthy `ControllerLike` trait that does not contain any code relying on runtime DI ;
- define a `I18nSupport` trait that requires a `MessagesApi` and provides convenient method to deal with langs and HTTP requests and responses ;
- remove the current behavior that makes it possible to use the `Messages` API without actually defining messages ;
- existing code base can migrate by adding an `import play.api.i18n.Mesages.Implicits._` which brings a `Messages` or a `MessagesApi` as long as there is an implicit `Application` in the scope.

**Comment by jroper** [comment]
Hey @julienrf,

I think changing the form helpers to accept an implicit `Messages` instead of `Lang` is good.  But a few comments:

* I don't think Java developers should have to worry about this in Play 2.4 (Play 3.0 will be a different question, we need to think hard about what to do there, eg we could switch to FreeMarker for Java devs or something like that, and have an injectable FreeMarker engine which provides the messages API in the context - but this is a whole different discussion to be had).  Requiring Java developers to add implicit parameters to their templates should be avoided, since for them this is a very confusing feature of Scala that most won't understand, making Scala templates even more foreign to them.  Rather, we have the `PlayMagicForJava` object, which, when the `PlayJava` plugin is enabled, is imported in templates (it's also available in all the templates under the `javaGuide` section of the documentation).  I think that should provide a global state based implicit messages.
* Is it that bad to have the `request2Lang` implicit in scope for people that don't use it?  Couldn't users just extend `Controller with I18nSupport`?
* There is a default implicit lang, but there's no default implicit messages.  Perhaps we should make it simple to provide one (that depends on global state), eg by just adding a template import. This would assist people with many existing templates that don't do i18n but make use of the form helpers, so they don't need to thread implicit messages through all the their templates if they don't want.
* We'll need to comprehensively document this in the migration guide.  Included in these docs will be:
 - Form helpers now take implicit messages
 - There is no default implicit messages (there is a default implicit lang) so now you need to always have implicit messages parameters in your templates (or use the work around mentioned above).
 - How to provide a messages instance from your controller (ie use I18nSupport)
* Is it necessary to change the `Messages.apply` object methods to take a `Messages` instance?  Do any of your other changes require this?  These methods are all going to be deprecated anyway, so I don't see any need to change their signature, it's nice to keep them around so that existing uses of them don't need to be changed.  For transitional support for users that change their templates to take an implicit `Messages` instead of implicit `Lang`, we could provide an implicit `messages2Lang` conversion.

**Comment by julienrf** [comment]
> Is it that bad to have the request2Lang implicit in scope for people that don't use it? Couldn't users just extend Controller with I18nSupport?

The problem I want to avoid is that user code may easily call this implicit conversion but the conversion is not safe (it relies on a runtime injected application). The idea of `ControllerLike` is to not rely at all on runtime injection.

**Comment by julienrf** [comment]
> There is a default implicit lang, but there's no default implicit messages. Perhaps we should make it simple to provide one (that depends on global state), eg by just adding a template import. This would assist people with many existing templates that don't do i18n but make use of the form helpers, so they don't need to thread implicit messages through all the their templates if they don't want.

There is one:

```scala
import play.api.i18n.Messages._ // brings an implicit `Messages`, provided you have an implicit `Application` in the scope
```

**Comment by julienrf** [comment]
> Is it necessary to change the Messages.apply object methods to take a Messages instance? Do any of your other changes require this? These methods are all going to be deprecated anyway, so I don't see any need to change their signature, it's nice to keep them around so that existing uses of them don't need to be changed. For transitional support for users that change their templates to take an implicit Messages instead of implicit Lang, we could provide an implicit messages2Lang conversion.

This is not “necessary” but it gives a better API to users: once they have an implicit `Messages` in scope, they can call `Messages("some.key")` to get its translation. Otherwise, the developer experience would look like [the following](https://github.com/julienrf/pfe-samples/blob/play-2.4.0-compile-time-di/pfe-scala/app/controllers/Authentication.scala#L32).

**Comment by julienrf** [comment]
> I don't think Java developers should have to worry about this in Play 2.4 […]. Requiring Java developers to add implicit parameters to their templates should be avoided […]. Rather, we have the `PlayMagicForJava` object, which, when the PlayJava plugin is enabled, is imported in templates (it's also available in all the templates under the javaGuide section of the documentation). I think that should provide a global state based implicit messages.

The problem is then that today the `Messages` API or the HTML helpers can be used without a “current” HTTP request. If you want to provide an implicit `Messages` API that reads the current HTTP context then the current usage is not possible anymore.

**Comment by julienrf** [comment]
Hi James, I added a commit that fixes the problem with Java users: they don’t need to add an implicit `Messages` parameter to their templates as long as they have a [running application](https://github.com/playframework/playframework/pull/3782/files#diff-b415c8bcfe76f7bb251a05568560294bR25) when they evaluate the template.

**Comment by julienrf** [comment]
Hi, I finally removed the `ControllerLike` trait.

I think the last point that remains is documentation. WDYT?

**Comment by julienrf** [comment]
I’ve just added some documentation and migration notes.

**Comment by jroper** [comment]
Hey @julienrf, just letting you know that I haven't forgotten this, just haven't had the time to fully digest what impact this is going to have on existing apps.

File: framework/src/play/src/main/scala/play/api/mvc/Results.scala [hunk 1]
```diff
@@ -251,34 +250,6 @@ case class Result(header: ResponseHeader, body: Enumerator[Array[Byte]],
   def removingFromSession(keys: String*)(implicit request: RequestHeader): Result =
     withSession(new Session(session.data -- keys))
 
-  /**
```
Comment (cchantep) [inline framework/src/play/src/main/scala/play/api/mvc/Results.scala]: Keeping those one as deprecated maybe?

File: framework/src/play/src/main/scala/play/api/mvc/Results.scala [hunk 2]
```diff
@@ -331,8 +302,49 @@ object Codec {
...
+     * For example:
+     * {{{
+     * implicit val lang = Lang("fr-FR")
+     * Ok(Messages("hello.world")).withLang(lang)
+     * }}}
+     *
+     * @param lang the language to store for the user
+     * @return the new result
+     */
+    def withLang(lang: Lang): Result =
```
Comment (cchantep) [inline framework/src/play/src/main/scala/play/api/mvc/Results.scala]: Add `@deprecated` ?

**Comment by cchantep** [comment]
@play-pull-request-validator

**Comment by cchantep** [comment]
A rebase may be useful

File: framework/src/play/src/main/scala/play/api/mvc/Results.scala (hunk 2 above)
Comment (julienrf) [inline framework/src/play/src/main/scala/play/api/mvc/Results.scala]: There would be a lot of things to deprecate. I remember a discussion with @jroper saying that maybe we will have a 2.5 version where we would effectively deprecate the old APIs but the plan is to make the migration path from 2.3 to 2.4 as simple as possible.

What do you think?

**Comment by cchantep** [comment]
At first for 2.4, I would keep the deprecated functions (in legacy trait or just where they are), with the annotation.

In this way, existing app would get deprecation warnings on update and the migration would be easy.

Then deprecated code could be cleaned up after.

**Comment by julienrf** [comment]
I rebased the commits.

**Comment by cchantep** [comment]
Please squash the commit (at least all except the last doc one).

**Comment by jroper** [comment]
@julienrf I'm happy with this, just squash the commits and we can merge.

**Comment by julienrf** [comment]
Hi guys, I squashed the commits

**Comment by cchantep** [comment]
Lgtm

**Comment by jroper** [comment]
Thanks @julienrf.

**Comment by slorber** [comment]
nice

In older play version I had to copy the following code:

```scala
  implicit def lang(implicit request: RequestHeader) = {
    play.api.Play.maybeApplication.map { implicit app =>
      val maybeLangFromCookie = request.cookies.get(Play.langCookieName).flatMap(c => Lang.get(c.value))
      maybeLangFromCookie.getOrElse(play.api.i18n.Lang.preferred(request.acceptLanguages))
    }.getOrElse(request.acceptLanguages.headOption.getOrElse(play.api.i18n.Lang.defaultLang))
  }
```

**Pull request merged**